ublox_derive/src/output.rs

```diff
@@ -0,0 +1,572 @@
+            }
+        }
+    }
+
+    if known_flags.len() != bitflags.consts.len() {
+        let user_flags: HashSet<_> = bitflags.consts.iter().map(|x| x.value).collect();
+        let set = user_flags.difference(&known_flags);
+        return Err(syn::Error::new(
+            bitflags.name.span(),
+            format!("Strange flags, not power of 2?: {:?}", set),
+        ));
+    }
+
+    let vis = &bitflags.vis;
+    let attrs = &bitflags.attrs;
+    let name = &bitflags.name;
+
+    let from = match bitflags.from_fn {
+        None => quote! {},
+        Some(UbxTypeFromFn::From) => quote! {
+            impl #name {
+                const fn from(x: #repr_ty) -> Self {
+                    Self::from_bits_truncate(x)
+                }
+            }
+        },
+        Some(UbxTypeFromFn::FromUnchecked) => unimplemented!(),
+    };
+
+    let into = match bitflags.into_fn {
+        None => quote! {},
+        Some(UbxTypeIntoFn::Raw) => quote! {
+            impl #name {
+                const fn into_raw(self) -> #repr_ty {
+                    self.bits()
+                }
+            }
+        },
+    };
+
+    Ok(quote! {
+        bitflags! {
+            #(#attrs)*
+            #vis struct #name : #repr_ty {
+                #(#items);*;
+            }
+        }
+        #from
+        #into
+    })
+}
+
+pub fn generate_code_for_parse(recv_packs: &RecvPackets) -> TokenStream {
+    let mut pack_enum_variants = vec![];
+    let mut matches = vec![];
+    let mut max_16_chain = None;
+
+    for name in &recv_packs.all_packets {
+        let ref_name = format_ident!("{}Ref", name);
+        pack_enum_variants.push(quote! {
+            #name(#ref_name <'a>)
+        });
+
+        matches.push(quote! {
+            (#name::CLASS, #name::ID) if <#ref_name>::validate(payload).is_ok()  => {
+                Ok(PacketRef::#name(#ref_name(payload)))
+            }
+        });
+        if let Some(prev_max_16_chain) = max_16_chain {
+            max_16_chain = Some(quote! { max_u16(#name::MAX_PAYLOAD_LEN, #prev_max_16_chain) });
+        } else {
+            max_16_chain = Some(quote! { #name::MAX_PAYLOAD_LEN });
+        }
+    }
+
+    let union_enum_name = &recv_packs.union_enum_name;
+    let unknown_var = &recv_packs.unknown_ty;
+    quote! {
+        #[doc = "All possible packets enum"]
+        pub enum #union_enum_name<'a> {
+            #(#pack_enum_variants),*,
+            Unknown(#unknown_var<'a>)
+        }
+
+        pub(crate) fn match_packet(class: u8, msg_id: u8, payload: &[u8]) -> Result<PacketRef, ParserError> {
+            match (class, msg_id) {
+                #(#matches)*
+                _ => Ok(PacketRef::Unknown(#unknown_var {
+                    payload,
+                    class,
+                    msg_id
+                })),
+            }
+        }
+
+        const fn max_u16(a: u16, b: u16) -> u16 {
+            [a, b][(a < b) as usize]
```

Comment:
I think it would just be clearer to do:
```
if b > a { b } else { a }
```

Reply:
Also, can we just use std::cmp::max?

Reply:
I don't get it. It is impossible to use `if` in `const fn`, and `std::cmp::max` is not `const fn`.

Reply:
Ah, I did not know that. Weird that `if` isn't allowed, but okay.

Reply:
For example you can use `if` inside C++ `constexpr` only starting from C++14, C++11 doesn't allow this. In fact `if` and `match` is implemented for `const fn`, but it is nigtly only, behind "feature gate" rust-lang/rust#49146